### DISCUSSION TITLE
`EMPTY-BUFFERS` effect on `BLK` ?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#161] [General]
Should `EMPTY-BUFFERS` set contents of `BLK` to zero?  If all the blocks are unassigned by `EMPTY-BUFFERS` then there is no current block being interpreted; should `BLK` reflect that condition?

**Comment by ruv** [comment]
> Should EMPTY-BUFFERS set contents of BLK to zero?

It should not. Because _formally_ the block will remain the input buffer anyway.  Then, it's incorrect to invoke `EMPTY-BUFFERS` when `BLK` is not zero. Then, an implicit ambiguous condition exists in this regard. Then, `EMPTY-BUFFERS` may throw an exception in this case.

Probably, it should be declared explicitly: an ambiguous condition exists if `BLK` contains a non-zero when `EMPTY-BUFFERS` is invoked.

**Reply by ruv** [reply]
I have recorded this [issue on forth-standard](https://forth-standard.org/standard/block/EMPTY-BUFFERS#reply-1295) too.

When a value in `blk` is not zero, this value is the number of a block being interpreted (by `load`), consequently, that block is assigned a block buffer, and that block buffer is the _input buffer_.

If `empty-buffers` is allowed to **free** buffers (free the memory that the buffers occupy), then it should not be executed if `blk` contains a nonzero value. Then, `empty-buffers` might throw an exception if `blk` contains a nonzero value.

**Reply by SirWumpus** [reply]
Is there a specific exception code that should be thrown, such as `-21 unsupported operation`?

**Reply by ruv** [reply]
The code `-21` is not suitable in this case. Since the problem is not in the capabilities of the system or environment, but in the system's state that is not compatible with the requested operation.

I don't see a suitable standard code.  A Forth system might use a system-defined code from the range {-4095...-256}. Or use something like `blk @ abort" Cannot empty-buffers while a block is loading"`.

**Reply by forthy42** [reply]
I see no problems with `empty-buffers` in the middle of interpreting a block.  The next `source` will call `block` again, and reload that block.

**Reply by ruv** [reply]
>  The next `source` will call `block` again, and reload that block.

At the first, `source` is not allowed to call `block`, I think. Since it may invalidate some block buffer addresses.

At the second, `source` shall return the same stack parameters till execution of `refill`, `restore-input`, or the input source is changed (all such cases are specified). But there is no guarantee that `block` (with the same parameter) returns the same address before and after `empty-buffers` .

The only way I see to allow `empty-buffers` during `load` is that `load` copies the block content into a separate buffer before interpret.  I still don't sure that the standard requires that.

**Reply by forthy42** [reply]
If you load from blocks, `source` has to and will call `block`.  The input buffer is the block referred to in the variable `blk`, and that block is obtained by `blk @ block` (and no, there's no copy made).

In Gforth, `block` gives you the same address for the same block number, that's however an implementation detail.  The currently loaded block isn't locked, though, so when you load another block that wants to use the same buffer space, there'll be problems if you expect the input buffer to survive other block I/O without calling `source` again.

**Reply by ruv** [reply]
> If you load from blocks, `source` has to and will call `block`.

Even if this is allowed, this is an implementation detail, isn't it?  For example, nothing in the standard prevents a system to copy a block content into a separated input buffer (by `load`), and then `source` might not call `block`.

> In Gforth, `block` gives you the same address for the same block number.

The following phrase (test 1):
```forth
:noname s" empty-buffers 123 . " 1 buffer dup 1024 blank swap cmove 1 load ; execute
```
does not print anything in Gforth, but prints "123" in some other systems.

Without `empty-buffers` it prints "123" in Gforth too:
```forth
:noname s" 123 . " 1 buffer dup 1024 blank swap cmove 1 load ; execute
```
So, providing the same address for the same block number does not solve the problem.

> there'll be problems if you expect the input buffer to survive other block I/O without calling `source` again

Do you mean that the following phrase in some block:
```forth
:noname source >in @ over >in ! /string evaluate ; execute
\ ... further block contents that performs block I/O during loading,
\ e.g. `load` or `thru`, but not `refill`
```
is ambiguous, even if the further block contents is unambiguous?

I think, the input buffer (whether it's a block buffer, or separate buffer), **shall be locked** to prevent it from being reused by another block until interpretation of the input buffer is completed.

I don't see anything in the standard that disallows a Forth system to lock a block buffer in this manner.


In the same time, [7.6.2.1330 `EMPTY-BUFFERS`](https://forth-standard.org/standard/block/EMPTY-BUFFERS) says:
- Unassign all block buffers. Do not transfer the contents of any [`UPDATE`](https://forth-standard.org/standard/block/UPDATE)d block buffer to mass storage.

An unassigned block buffer can be assigned to any block. A locked block buffer cannot be assigned to any other block, and shall not be reloaded from mass storage (since they can be in modified state). It means, **a locked block buffer cannot be unassigned**.

Obviously, the behavior of `empty-buffers` in this case is insufficiently specified (and the test 1 also shows that).  Therefore, **an ambiguous condition exists**, if `empty-buffers` is executed when loading of some block buffer is not completed. NB: referring `blk` is not enough, because `empty-buffers` can be executed from a nested `evaluate` or `included`.

**Reply by ruv** [reply]
It seems, the standard allows the system to reuse the same input buffer in a nested input source. The section [3.3.3.5 Input buffers](https://forth-standard.org/standard/usage#usage:inbuf) says:
> The address and length returned by [`SOURCE`](https://forth-standard.org/standard/core/SOURCE), the string returned by [`PARSE`](https://forth-standard.org/standard/core/PARSE), and directly computed input-buffer addresses **are valid only until** the text interpreter does I/O to refill the input buffer or **the input source is changed**.

(emphasis mine)

This means that the standard allows a Forth system to print both `-1` (the strings are equal) and `0` (the strings are not equal) when interpreting the phrase:
```forth
parse-name foo  s" ( sd -- flag ) parse-name foo compare 0=" evaluate .
```
Instead of `evaluate` an equivalent `included` or `load` can be used.

If this reading is correct, a consequence of this is that a string from the input buffer cannot be passed into a nested invocation of the Forth text interpreter (on different input source).

Do you know Forth systems that print `0` in the above test with some combination of input source nesting?

-----

It's unclear whether a Forth system is allowed to use different memory ranges for the input buffer before and after of a nested invocation of the Forth text interpreter.

If it's allowed, then the following phrase may print `0` (i.e. that the strings are not equal) in some combination of input source nesting:
```forth
parse-name foo  s" ( -- ) " evaluate  parse-name foo compare 0= .
```
Moreover, in this case it's an incorrect program since the string of the first "foo" may become invalid after `evaluate` (as well as `included` or `load`).

Do you know any Forth systems where the address returned by `source` is different before and after nesting+unnesting of the input source, in some combination of input source kinds?

**Reply by gerryjackson** [reply]
Have you see the Technical Committee's reply to RFI 0006 that contains the following towards the end of the discussion:
"... both the letter of the standard and the intent of the TC are that **EVALUATE** strings be processed in place. ... the Standard clearly states that the **EVALUATE** string **is** the 'input buffer' ..."

You might disagree with the the word 'clearly'.

**Reply by ruv** [reply]
@gerryjackson, thank you!  I was just writing an update to my post in this regard.

According to the [testcase](https://forth-standard.org/standard/core/EVALUATE#contribution-17), in case of `evaluate` the standard guarantees that the input buffer is the same (before and after), and it may be passed to a nested invocation of the Forth text interpreter.

Then my questions remain relevant only for files, blocks, and the user input device (it cannot be nested, but can be the initial input source).

**Reply by forthy42** [reply]
Historically, Forth systems didn't have ample of RAM.  In-place zero copy evaluation therefore is what's expected.  Using one single buffer for all file inputs is possible, too (though most file-system based Forths have enough memory and don't need this); just for terminal input, you can't reuse the buffer, because you have no way to reread the same input again.

I would suggest to read the “make block u the input source” literal: `SOURCE` is equal to `BLK @ BLOCK B/BLK`.  The “encompass” in the text is best interpreted this way.  `BLK @ IF  BLK @ BLOCK ELSE TIB @ THEN` is what you find in fig-Forth where SOURCE should have been factored out.  Screen 47:

https://github.com/ForthHub/FIG-Forth/blob/master/fig.fth

**Reply by SirWumpus** [reply]
Forth draft 19-1 section 3.3.3.4 Text-literal regions
> A system must provide at least two transient buffers for use with C", S" and S\" strings. 

So `PARSE-NAME` uses the input buffer and `S"` uses a transient buffer in interpretation state.  So the test should always return TRUE (-1)
```
parse-name foo  s" ( sd -- flag ) parse-name foo compare 0=" evaluate .
```

Section A.11.6.1.1717 INCLUDE-FILE proposes two methods methods for input; the first case would reuse the same input buffer, the second would use different line buffers.

> 2) Allocate a separate line buffer for each active text input file, using that buffer as the input
buffer. This method avoids the “seek and reread” step, and allows the use of “pseudo-files”
such as pipes and other sequential-access-only communication channels

For a hosted Forth, I suspect the second method is used and the suggested test again returns TRUE.

**Reply by ruv** [reply]
> So `PARSE-NAME` uses the input buffer and `S"` uses a transient buffer in interpretation state. So the test should always return `TRUE`

Yes, in case of `evaluate` the test is always `true`. 
But what if this string will be a block content interpreted by `load` (instead of `evaluate`), or a file content interpreted by `included`?


> ```forth
> parse-name foo  s" ( sd -- flag ) parse-name foo compare 0=" evaluate .
> ```

>   2. Allocate a separate line buffer for each active text input file,

> For a hosted Forth, I suspect the second method is used and the suggested test again returns TRUE.

Agreed.
The only question is: do we know of any Forth system (claimed to be at least a Forth-94 standard system) that uses the first method, or some another method that invalidates the address and/or content of the input buffer?

**Comment by MitchBradley** [comment]
When I started doing Forth in 1982, there was a slight justification for using blocks on 8080 machines with 5.25" floppy drives, but within a few years it became so much easier to use files.  By the time the standard was published in 1994, even most of the old-school die-hards were using files.  We kept blocks in the standard just because a few people were nostalgic.  Now, 30 years later, with a Moore's Law exponent of 10, it is hard to imagine that anyone still cares about the BLOCK subsystem.

**Reply by SirWumpus** [reply]
I think the BLOCK subsystem (even as a block file) is cool, easy to implement quickly, a smaller word set than the File Access, and probably still of interest for embedded systems.

**Reply by SirWumpus** [reply]
The only thing I wish was possibly was to have some means to set the block size; maybe define the block size in the first cell of block 1 or by LOADing block 1, but that's just a minor thing.  Having 4K or 8k blocks might be more useful for modern mass storage.

**Reply by SirWumpus** [reply]
For example
```forth
BLK-SIZE @ 1024 = IF 8096 BLK-SIZE ! 1 LOAD THEN

( continue using the extended block size)
```

**Reply by MitchBradley** [reply]
I do a lot of embedded work on ESP32 MCUs that cost less than $2 and have a minimum of 8MB of FLASH storage that can be managed as files via a LittleFS filesystem that is well-supported by the SDK.  I guess if you are writing a bare-metal Forth from scratch, you could get something working with BLOCK with less code, but probably not more quickly compared to using the existing SDKs. Bare-metal was useful to me 15 years ago when I did the Forth firmware for the One Laptop Per Child computer and we needed to have control over every line of code for security considerations, but the number of people who do stuff like that anymore is very close to zero.  Even then, it used files, because I had implemented open source Forth code for multiple standard file systems long before.

**Reply by ruv** [reply]
> I think the BLOCK subsystem (even as a block file) is cool, easy to implement quickly, a smaller word set than the File Access, and probably still of interest for embedded systems.

Blocks seem to map easily to sectors of a mass storage device. Therefore, some simple file system can be implemented as a standard program using the Block word set. Maybe just like a toy.
